### ISSUE TITLE
Health check sometimes stops responding in EKS cluster

### ISSUE THREAD

**Issue by bingzhang24** [kubernetes-sigs/kwok#950]
### How to use it?

- [X] kwok
- [ ] kwokctl --runtime=docker (default runtime)
- [ ] kwokctl --runtime=binary
- [ ] kwokctl --runtime=nerdctl
- [ ] kwokctl --runtime=kind

### What happened?

We have kwok deployed in an EKS cluster using the default manfiest "https://github.com/kubernetes-sigs/kwok/releases/download/v0.4.0/kwok.yaml".  We use the default node stages in the repository [node-fast.yaml](stages/node-fast.yaml) and [node-heartbeat.yaml](stages/node-heartbeat.yaml). 

We observed that sometimes the Kubenetes health check would fail, causing the container to be restarted, but the issue would remain for the new container, thus causing the pod to stuck in a failure-restart loop.  The issue would only go away temporary if `kubectl rollout restart deploy` was used, which creates a new pod.  It happens more frequently after new kwok-managed nodes are created, but it can happen in other situations too. 

* Sometimes a TCP connection couldn't be made even inside the pod:

```
$ kubectl exec -n kube-system  -it kwok-controller-5bd7f448bb-strqz -- sh
$ nc -vz localhost 10247
nc: localhost ([::1]:10247): Operation timed out
```

* Sometimes `nc` was successful, but an HTTP request would lead to no response:

```
$ nc localhost 10247
GET /healthz HTTP/1.0
Host: localhost

# hangs. no output 
```

While such problems occured, the controller appears to be running OK. The log shows that it was processing node heartbeats and pods as expected. 

### What did you expect to happen?

Health check should not fail while the controller remains functional.

### How can we reproduce it (as minimally and precisely as possible)?

* Create an AWS EKS cluster
* Deploy kwok using "https://github.com/kubernetes-sigs/kwok/releases/download/v0.4.0/kwok.yaml".  
* Deploy node stages [node-fast.yaml](stages/node-fast.yaml) and [node-heartbeat.yaml](stages/node-heartbeat.yaml). 
* Create 8 nodes. (Actually number probably does not matter much, but it seems that more nodes trigger the problem more often.)
* Add 8 more nodes and wait for the liveness probe to fail. It can take up to one minute. 

### Anything else we need to know?

* Other HTTP endpoints would stop responding too, such as /metrics.
* We have not seen the issue in a local cluster (kind). 
* When the container was restarted due to liveness probe failure, the log shows the following error, which we believe is the result, not the cause of the restart. 
```
{"time":"2024-01-30T21:44:53.205626041Z","level":"ERROR","source":{"function":"sigs.k8s.io/kwok/pkg/kwok/cmd.runE.func1","file":"/workspace/pkg/kwok/cmd/root.go","line":348},"msg":"Failed to run server","err":"context canceled"}
```

### Kwok version

<details>

```console
$ kwok --version
kwok version v0.4.0 go1.20.5 (linux/amd64)
$ kwokctl --version
# N/A -- not used
```

</details>

### OS version

<details>

```console

$ cat /etc/os-release
NAME="Alpine Linux"
ID=alpine
VERSION_ID=3.18.3
PRETTY_NAME="Alpine Linux v3.18"
HOME_URL="https://alpinelinux.org/"
BUG_REPORT_URL="https://gitlab.alpinelinux.org/alpine/aports/-/issues"

$ uname -a
Linux kwok-controller-5bd7f448bb-strqz 5.10.192-183.736.amzn2.x86_64 #1 SMP Wed Sep 6 21:15:41 UTC 2023 x86_64 Linux
```
</details>

**Comment by Huang-Wei** [comment]
> Create an AWS EKS cluster

Bing, which EKS version are you using? and how many (real) nodes were provisioned?

**Comment by bingzhang24** [comment]
This is EKS version 1.28. There are 42 real nodes.

**Comment by wzshiming** [comment]
😄 This is a known bug and the fix is included in 0.5, which will be released today

**Comment by bingzhang24** [comment]
What is the bug and the PR that fixed it?

**Comment by Huang-Wei** [comment]
Fixed by #924

**Comment by wzshiming** [comment]
[v0.5.1](https://github.com/kubernetes-sigs/kwok/releases/tag/v0.5.1) has been released, fixing all known bugs.

/close
